File: android/src/main/java/uk/co/playerdata/reactnativemcumanager/DeviceUpdate.java
package uk.co.playerdata.reactnativemcumanager;

import android.bluetooth.BluetoothDevice;
import android.net.Uri;
import android.util.Log;

import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.WritableMap;

import java.io.IOException;
import java.io.InputStream;

import io.runtime.mcumgr.McuMgrTransport;
import io.runtime.mcumgr.ble.McuMgrBleTransport;
import io.runtime.mcumgr.dfu.FirmwareUpgradeCallback;
import io.runtime.mcumgr.dfu.FirmwareUpgradeController;
import io.runtime.mcumgr.dfu.FirmwareUpgradeManager;
import io.runtime.mcumgr.exception.McuMgrException;

public class DeviceUpdate implements FirmwareUpgradeCallback {
    private final String TAG = "DeviceUpdate";
    private final BluetoothDevice device;
    private final Promise promise;
    private final ReactApplicationContext context;
    private final Uri updateFileUri;
    private final McuManagerModule manager;
    private FirmwareUpgradeManager dfuManager;
    private int LastNotification = -1;
    private McuMgrTransport transport;
    private Boolean noFailures = true;

    public DeviceUpdate(BluetoothDevice device, Promise promise, ReactApplicationContext context, Uri updateFileUri, McuManagerModule manager) {
        this.device = device;
        this.promise = promise;
        this.context = context;
        this.updateFileUri = updateFileUri;
        this.manager = manager;
    }

    public void startUpdate() {
        doUpdate(updateFileUri);
    }

    public void cancel() {
        this.dfuManager.cancel();
        this.disconnectDevice();
        this.promise.reject("Update cancelled");
        this.manager.unsetUpdate();
    }

    public void disconnectDevice() {
        if (this.transport != null) {
            this.transport.release();
        }
    }

    private void doUpdate(Uri updateBundleUri) {
        this.transport = new McuMgrBleTransport(this.context, device);
        this.dfuManager = new FirmwareUpgradeManager(this.transport, this);

        try {
            InputStream stream = context.getContentResolver().openInputStream(updateBundleUri);

            byte[] imageData = new byte[(int) stream.available()];
            stream.read(imageData);

            this.dfuManager.setMode(FirmwareUpgradeManager.Mode.TEST_AND_CONFIRM);
            this.dfuManager.start(imageData);
        } catch (IOException e) {
            e.printStackTrace();
            this.disconnectDevice();
            this.manager.unsetUpdate();
            this.promise.reject(e);
        } catch (McuMgrException e) {
            e.printStackTrace();
            this.disconnectDevice();
            this.manager.unsetUpdate();
            this.promise.reject(e);
        }
    }


    @Override
    public void onUpgradeStarted(FirmwareUpgradeController controller) {

    }

    @Override
    public void onStateChanged(FirmwareUpgradeManager.State prevState, FirmwareUpgradeManager.State newState) {
        Log.i(TAG, "Leaving state: " + prevState);
        Log.i(TAG, "Entering state: " + newState);
        WritableMap stateMap = Arguments.createMap();

        stateMap.putString("bleId", this.device.getAddress());
        stateMap.putString("state", newState.name());
        this.manager.updateStateCB(stateMap);
    }

    @Override
    public void onUpgradeCompleted() {
        this.manager.unsetUpdate();
        this.disconnectDevice();
        this.promise.resolve(null);
    }

    @Override
    public void onUpgradeFailed(FirmwareUpgradeManager.State state, McuMgrException error) {
        if (state == FirmwareUpgradeManager.State.RESET && noFailures) {
            Log.w(TAG, "experienced first time failure in reset state");
            //assume the device has taken slightly longer to come back up and has dropped bluetooth connection the first time this happens
            noFailures = false;
            this.disconnectDevice();
            try {
                Thread.sleep(4000);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
            doUpdate(updateFileUri);
            return;
        }
        Log.e(TAG,"failed while in state: " + state);
        Log.e(TAG,"with error: " + error);
        this.manager.unsetUpdate();
        this.disconnectDevice();
        this.promise.reject(error);
    }

    @Override
    public void onUpgradeCanceled(FirmwareUpgradeManager.State state) {
        this.manager.unsetUpdate();
        this.disconnectDevice();
        this.promise.reject("Update cancelled");
    }

    @Override
    public void onUploadProgressChanged(int bytesSent, int imageSize, long timestamp) {
        int progress_percent = bytesSent *100/imageSize;
        if (progress_percent != this.LastNotification) {
            this.LastNotification = progress_percent;
            WritableMap progressMap = Arguments.createMap();

            progressMap.putString("bleId", this.device.getAddress());
            progressMap.putString("progress", String.format("%d", progress_percent, 2));
            this.manager.updateProgressCB(progressMap);
        }
    }
}

File: android/src/main/java/uk/co/playerdata/reactnativemcumanager/DeviceUpdate.kt
package uk.co.playerdata.reactnativemcumanager

import android.bluetooth.BluetoothDevice
import android.net.Uri
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReadableMap
import io.runtime.mcumgr.ble.McuMgrBleTransport
import io.runtime.mcumgr.dfu.FirmwareUpgradeCallback
import io.runtime.mcumgr.dfu.FirmwareUpgradeController
import io.runtime.mcumgr.dfu.FirmwareUpgradeManager
import io.runtime.mcumgr.exception.McuMgrException
import java.io.IOException

val UpgradeModes = mapOf(
    1 to FirmwareUpgradeManager.Mode.TEST_AND_CONFIRM,
    2 to FirmwareUpgradeManager.Mode.CONFIRM_ONLY,
    3 to FirmwareUpgradeManager.Mode.TEST_ONLY
)

class DeviceUpdate(
        private val device: BluetoothDevice, private val promise: Promise, private val context: ReactApplicationContext,
        private val updateFileUri: Uri, private val updateOptions: ReadableMap, private val manager: McuManagerModule
        ) : FirmwareUpgradeCallback {

    private var lastNotification = -1
    private var transport = McuMgrBleTransport(context, device)
    private var dfuManager = FirmwareUpgradeManager(transport, this)

    fun startUpdate() {
        doUpdate(updateFileUri)
    }

    fun cancel() {
        dfuManager.cancel()
        disconnectDevice()
        promise.reject("Update cancelled")
        manager.unsetUpdate()
    }

    private fun disconnectDevice() {
        transport.release()
    }

    private fun doUpdate(updateBundleUri: Uri) {
        val estimatedSwapTime = updateOptions.getInt("estimatedSwapTime") * 1000
        val modeInt = if (updateOptions.hasKey("upgradeMode"))  updateOptions.getInt("upgradeMode") else 1
        val upgradeMode = UpgradeModes[modeInt] ?: FirmwareUpgradeManager.Mode.TEST_AND_CONFIRM

        dfuManager.setEstimatedSwapTime(estimatedSwapTime)

        try {
            val stream = context.contentResolver.openInputStream(updateBundleUri)
            val imageData = ByteArray(stream!!.available())

            stream.read(imageData)

            dfuManager.setMode(upgradeMode)
            dfuManager.start(imageData)
        } catch (e: IOException) {
            e.printStackTrace()
            disconnectDevice()
            manager.unsetUpdate()
            promise.reject(e)
        } catch (e: McuMgrException) {
            e.printStackTrace()
            disconnectDevice()
            manager.unsetUpdate()
            promise.reject(e)
        }
    }

    override fun onUpgradeStarted(controller: FirmwareUpgradeController) {}

    override fun onStateChanged(prevState: FirmwareUpgradeManager.State, newState: FirmwareUpgradeManager.State) {
        val stateMap = Arguments.createMap()
        stateMap.putString("bleId", device.address)
        stateMap.putString("state", newState.name)
        manager.updateStateCB(stateMap)
    }

    override fun onUpgradeCompleted() {
        manager.unsetUpdate()
        disconnectDevice()
        promise.resolve(null)
    }

    override fun onUpgradeFailed(state: FirmwareUpgradeManager.State, error: McuMgrException) {
        manager.unsetUpdate()
        disconnectDevice()
        promise.reject(error)
    }

    override fun onUpgradeCanceled(state: FirmwareUpgradeManager.State) {
        manager.unsetUpdate()
        disconnectDevice()
        promise.reject("Update cancelled")
    }

    override fun onUploadProgressChanged(bytesSent: Int, imageSize: Int, timestamp: Long) {
        val progressPercent = bytesSent * 100 / imageSize
        if (progressPercent != lastNotification) {
            lastNotification = progressPercent
            val progressMap = Arguments.createMap()
            progressMap.putString("bleId", device.address)
            progressMap.putString("progress", String.format("%d", progressPercent, 2))
            manager.updateProgressCB(progressMap)
        }
    }
}
